stdlib/src/builtin/string.mojo

```diff
@@ -1405,36 +1408,109 @@ struct String(
             substr._strref_dangerous(), start=start
         )
 
-    fn split(self, delimiter: String) raises -> List[String]:
-        """Split the string by a delimiter.
+    fn isspace(self) -> Bool:
+        """Determines whether the given String is a python
+        whitespace String. This corresponds to Python's
+        [universal separators](
+            https://docs.python.org/3/library/stdtypes.html#str.splitlines)
+        `" \\t\\n\\r\\f\\v\\x1c\\x1e\\x85\\u2028\\u2029"`.
+
+        Returns:
+            True if the String is one of the whitespace characters
+                listed above, otherwise False."""
+        # TODO add line and paragraph separator as stringliteral
+        # once unicode escape secuences are accepted
+        # Unicode Line Separator: \ u2028
+        # Unicode Paragraph Separator: \ u2029
+        return self in List[String](
+            String(" "),
+            String("\t"),
+            String("\n"),
+            String("\r"),
+            String("\v"),
+            String("\f"),
+            String("\x1c"),
+            String("\x1e"),
+            String("\x85"),
+            String(List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x38)),
+            String(List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x39)),
+        )
+
+    fn split(self, sep: String = "", maxsplit: Int = -1) -> List[String]:
```

Comment:
The behavior of returning an empty string for `sep=""` is kind of strange and actually breaks an invariant: `sep.join(some_str.split(sep)) == some_str` should be true for any `sep` and any `some_str`. Python simply rejects empty separators (with `ValueError: empty separator`), and we could certainly do this. The (IMO better) alternative is to return a list of the characters. WDYT?

Reply:
You can check this by running `"   ".split(" ")` and `"   ".split()` in python.

Reply:
> to return a list of the characters. WDYT?
I think it makes a lot of sense, though it adds yet another branch to the code that could be parametrized... but yeah it's inevitable. I hadn't actually tested the `.split("")` with python so thanks for the save.

Reply:
`split("")` shouldn't be possible IMO, since there's an infinite number of empty strings between every symbol. If you want to get a list of the characters of a string I would opt for `list(some_string)`

Reply:
I understand but how would we deal with it then? IMO we shouldn't raise in a func that is just splitting, the other option is to return an empty list, or a list with the whole string as the first and only item. And I do like the idea of returning chars since it would also provide an API for splitting a string into its chars, instead of iterating and building them or using `List[String](String.as_bytes())` (I'm not sure this implicit casting even works or if the user would have to build a String for each item in the byte List)

Reply:
I'm on board with returning a list of the characters from `.split("")`. I'm not typically a fan of deviating from Python behaviour but it's not really a useful safeguard to begin with and it's not worth "colouring" this method with `raises` to avoid an altered behaviour from exactly one possible input.

Reply:
Splitting on an empty string is ambiguous and conceptually wrong. If the goal is to be a superset of Python then I don't see why we should change Python's behaviour (unless Python changes it too).

As a fix we could add an overload like:
```mojo
fn split[sep: StringRef](maxsplit=-1):
    constrained[sep != "", "empty separator"]()
```

Reply:
I gave this some thought, and I think I mostly agree with @siitron and @bgreni: splitting on an empty string is ambiguous, and we shouldn't risk unexpected behavior by allowing it (even when it might feel natural to some of us). Let's keep it simple, and make this function raising. If this turns out to be a perf problem, there are ways we can deal with it, but our default should be following Python and providing safe, unsurprising APIs.